File: 2021 KAKAO BLIND RECRUITMENT/합승 택시 요금.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 987654321;

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int adj[n+1][n+1];
    for(int i = 0; i <= n; ++i)
        for(int j = 0; j <= n; ++j) {
        	if(i == j) adj[i][j] = 0;	    
            else adj[i][j] = INF;
        }
    for(int i = 0; i < fares.size(); ++i) {
        int c = fares[i][0];
        int d = fares[i][1];
        int f = fares[i][2];
      	adj[c][d] = f;
        adj[d][c] = f;
    }
    for(int k = 1; k <= n; ++k) 
        for(int i = 1; i <= n; ++i) 
            for(int j = 1; j <= n; ++j) 
                adj[i][j] = min(adj[i][j], adj[i][k] + adj[k][j]);
    int answer = INF;
    for(int i = 1; i <= n; ++i) {
        if(adj[s][i] == INF || adj[i][a] == INF || adj[i][b] == INF) continue;
        int cost = adj[s][i];
        cost += adj[i][a];
        cost += adj[i][b];
        answer = min(answer, cost);
    }
    return answer;
}

File: 2021 KAKAO BLIND RECRUITMENT/합승 택시 요금.java
class Solution {
    static final int INF = 40000000;
    int[][] Dist = new int[200][200];
    
    void floyd(int n) {
        for (int k = 0; k < n; ++k) {
            for (int i = 0; i < n; ++i) {
                for (int j = 0; j < n; ++j) {
                    Dist[i][j] = Math.min(Dist[i][j], Dist[i][k] + Dist[k][j]);
                }
            }
        }
    }
    
    public int solution(int n, int s, int a, int b, int[][] fares) {
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (i == j)
                    Dist[i][j] = 0;
                else
                    Dist[i][j] = INF;
            }
        }
        
        for (int[] edge : fares) {
            Dist[edge[0]-1][edge[1]-1] = edge[2];
            Dist[edge[1]-1][edge[0]-1] = edge[2];
        }
        
        floyd(n);
        
        --s;
        --a;
        --b;
        int answer = INF;
        for (int k = 0; k < n; ++k) {
            answer = Math.min(answer, Dist[s][k] + Dist[k][a] + Dist[k][b]);
        }
        return answer;
    }
}
